core/module.cc

```diff
@@ -47,38 +34,73 @@ task_id_t task_to_tid(struct task *t) {
   return INVALID_TASK_ID;
 }
 
-int num_module_tasks(Module *m) {
-  int cnt = 0;
+// -------------------------------------------------------------------------
 
-  for (task_id_t id = 0; id < MAX_TASKS_PER_MODULE; id++)
-    if (m->tasks[id]) cnt++;
+Module *ModuleBuilder::CreateModule(const std::string &name) const {
+  Module *m = module_generator_();
+  m->set_name(name);
+  m->set_module_builder(this);
+  return m;
+}
 
-  return cnt;
+bool ModuleBuilder::AddModule(Module *m) {
+  return all_modules_.insert({m->name(), m}).second;
 }
 
-size_t list_modules(const Module **p_arr, size_t arr_size, size_t offset) {
-  size_t ret = 0;
-  size_t iter_cnt = 0;
+int ModuleBuilder::DestroyModule(Module *m, bool erase) {
+  int ret;
+  m->Deinit();
 
-  struct ns_iter iter;
+  // disconnect from upstream modules.
+  for (int i = 0; i < m->igates.curr_size; i++) {
+    ret = m->DisconnectModulesUpstream(i);
+    if (ret) return ret;
+  }
 
-  ns_init_iterator(&iter, NS_TYPE_MODULE);
-  while (1) {
-    Module *module = (Module *)ns_next(&iter);
-    if (!module) break;
+  // disconnect downstream modules
+  for (gate_idx_t i = 0; i < m->ogates.curr_size; i++) {
+    ret = m->DisconnectModules(i);
+    if (ret) return ret;
+  }
 
-    if (iter_cnt++ < offset) continue;
+  m->DestroyAllTasks();
 
-    if (ret >= arr_size) break;
+  if (erase) all_modules_.erase(m->name());
 
-    p_arr[ret++] = module;
+  mem_free(m->ogates.arr);
+  mem_free(m->igates.arr);
+  delete m;
+  return 0;
+}
+
+void ModuleBuilder::DestroyAllModules() {
+  int ret;
+  for (auto it = all_modules_.begin(); it != all_modules_.end(); ++it) {
+    ret = DestroyModule(it->second, false);
+    if (ret) {
+      LOG(ERROR) << "Error destroying module '" << it->first << "' (errno = " << ret << ")";
+      ++it;
+      continue;
+    }
+    all_modules_.erase(it);
```

Comment:
Since you do the erase in this method, you could do a postincrement on the iterator.